docs/recursion.md

```diff
@@ -0,0 +1,118 @@
+---
+title: Recursion
+---
+
+There are three places recursion is commonly used: bindings, types, and modules.
+
+Feature  | Recursive           | Non-recursive
+---------|---------------------|--------------
+Bindings | `let rec fn = ...`  | `let fn = ...`
+Types    | `type t = ...`      | `type nonrec t = ...`
+Modules  | `module rec A ...`  | `module A ...`
+
+## Recursive Bindings
+
+By default the name of a binding is not available for use on the right side of
+that binding. This applies to all `let` bindings, including function
+definitions. This behavior is required for
+[Binding Shadowing](let-binding.md#binding-shadowing) to work:
+
+```reason
+let x = 10;
+/* If bindings were recursive by default this would form a cycle and break */
+let x = x + 10;
+```
+
+The natural way to write a recursive function will have an error:
+
+```reason
+let infiniteRecursion = () => {
+  /* Error: Unbound value infiniteRecursion */
+  infiniteRecursion();
+};
+```
+
+Opt-in to a recursive binding using the `rec` keyword:
+
+```reason
+let rec infiniteRecursion = () => {
+  infiniteRecursion();
+};
+```
+
+## Mutual Recursion
+
+Mutually recursive functions use the `and` keyword:
```

Comment:
It might be worth coming back to this to elaborate (maybe at the end)
`and` makes the bindings happen "simultaneously". If the leading binding is `rec`, it means the set of `and`ed bindings cannot see each other at all. If the leading binding is `rec` then it means all the bindings can see each other, and themselves.